classes/submission/PKPSubmissionFileDAO.inc.php

```diff
@@ -0,0 +1,343 @@
+<?php
+
+/**
+ * @file classes/submission/PKPSubmissionFileDAO.inc.php
+ *
+ * Copyright (c) 2014-2020 Simon Fraser University
+ * Copyright (c) 2003-2020 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class PKPSubmissionFileDAO
+ * @ingroup submission
+ * @see SubmissionFile
+ *
+ * @brief Operations for retrieving and modifying submission files
+ */
+use Illuminate\Database\Capsule\Manager as Capsule;
+
+import('lib.pkp.classes.db.SchemaDAO');
+import('lib.pkp.classes.submission.Genre'); // GENRE_CATEGORY_... constants
+import('lib.pkp.classes.plugins.PKPPubIdPluginDAO');
+import('lib.pkp.classes.submission.SubmissionFile');
+
+abstract class PKPSubmissionFileDAO extends SchemaDAO implements PKPPubIdPluginDAO {
+
+	/** @copydoc SchemaDAO::$schemaName */
+	public $schemaName = SCHEMA_SUBMISSION_FILE;
+
+	/** @copydoc SchemaDAO::$tableName */
+	public $tableName = 'submission_files';
+
+	/** @copydoc SchemaDAO::$settingsTableName */
+	public $settingsTableName = 'submission_file_settings';
+
+	/** @copydoc SchemaDAO::$primaryKeyColumn */
+	public $primaryKeyColumn = 'submission_file_id';
+
+	/** @copydoc SchemaDAO::$primaryTableColumns */
+	public $primaryTableColumns = [
+		'assocId' => 'assoc_id',
+		'assocType' => 'assoc_type',
+		'createdAt' => 'created_at',
+		'fileId' => 'file_id',
+		'fileStage' => 'file_stage',
+		'genreId' => 'genre_id',
+		'id' => 'submission_file_id',
+		'sourceSubmissionFileId' => 'source_submission_file_id',
+		'submissionId' => 'submission_id',
+		'updatedAt' => 'updated_at',
+		'uploaderUserId' => 'uploader_user_id',
+		'viewable' => 'viewable',
+	];
+
+	/**
+	 * Create a new DataObject of the appropriate class
+	 *
+	 * @return DataObject
+	 */
+	public function newDataObject() {
+		return new SubmissionFile();
+	}
+
+	/**
+	 * @copydoc SchemaDAO::_fromRow()
+	 */
+	public function _fromRow($primaryRow) {
+		$submissionFile = parent::_fromRow($primaryRow);
+
+		// Set the primary locale from the submission
+		$locale = Capsule::table('submissions as s')
```

Comment:
Couldn't this be done in the fetch that results in the `_fromRow` call? It seems needlessly expensive to do it here.

Reply:
Yes, I think this is an area where we could streamline things. The way the QueryBuilders are architected, though, the built query is used for getting data other than complete objects, for example [getting ids and counts](https://github.com/pkp/pkp-lib/pull/6292/files/31b98d3193211ef26c5c4255dea02ec0a8716e6a#diff-573f445e6ee526f514efb70857ad7d2a7a390a32cbf6b09cddf605e8e79c5591R157-R174).

When getting a collection of objects (`PKPSubmissionFileService::getMany(...)`), we get the same [query object](https://github.com/pkp/pkp-lib/pull/6292/files/31b98d3193211ef26c5c4255dea02ec0a8716e6a#diff-007b39807ba72de25a23e46d4ac9235dc5328d76fb97f461b45a47061785495dR55). So I think the select would need to be added in `PKPSubmissionFileService::getMany(...)`. Or, probably better in order to keep database details in the querybuilder/dao, a new method needs to be added to the query builder to get all the data needed to instantiate an object with `_fromRow()`. So instead of doing `new MyQueryBuilder()->getQuery()` we'd do something like `new MyQueryBuilder->getObjects()` which would include the `_fromRow` stuff?

Do you have a preference? (This is especially egregious in [PublicationDAO::_fromRow](https://github.com/pkp/pkp-lib/blob/master/classes/publication/PKPPublicationDAO.inc.php#L49-L78), although there maybe generators are the better approach since that data is not always used?

Reply:
The cost to selecting an extra column in cases where it's not needed (since `submissions` should already be joined) is close enough to zero that I think that trumps any conditional code to join it when needed. For complicated cases, we might need to add some methods to the DAO to expose the QueryBuilder object that's being constructed in the throwaway DAO instance in order to allow outside friendly code to augment it -- but I can't think of a case off the top of my head.

Reply:
(If you run into trouble with counting when columns are already selected, a quick-and-easy if not totally pretty solution might be to wrap a whole query in a `SELECT COUNT` -- much as I'm doing in the DAO rewrite for ADODB removal, but hopefully using a query builder. Apparently MySQL (and presumably PostgreSQL) are pretty quick to optimize this sensibly. https://github.com/asmecher/pkp-lib/blob/6381fdc689be77415bc712309c5ece5346f25e58/classes/db/DAO.inc.php#L134..L137)

Reply:
I think that this will be addressed when we rework the DAOs and QueryBuilders. In my experiments, we separate out the [selection critera](https://github.com/NateWr/pkp-lib/blob/daos_querybuilders/classes/announcement/DAO.inc.php#L73-L83) from the query building helper, and that should make it easy to optimize the selection queries differently whether we're getting full objects or just IDs or a count.

In the meantime, I've addressed this by writing the column selection and join into `PKPSubmissionFileService::getMany()`.

```
	public function getMany($args = null) {
		$submissionFileQO = $this
			->getQueryBuilder($args)
			->getQuery()
			->leftJoin('submissions as s', 's.submission_id', '=', 'sf.submission_id')
			->select(['sf.*', 's.locale as locale']);
		$submissionFileDao = DAORegistry::getDAO('SubmissionFileDAO');
		$result = $submissionFileDao->retrieve($submissionFileQO->toSql(), $submissionFileQO->getBindings());
		$queryResults = new DAOResultFactory($result, $submissionFileDao, '_fromRow');

		return $queryResults->toIterator();
	}
```